### PR TITLE
QA: Introduce feature_blocksign.py

### PR DESCRIPTION
**Pull request by jtimon**
Let's add this test that @stevenroose wrote, I only slightly modified it.

Later I want to extend it to test things other than 3 of 3, including a single signer (without using multisig if it's 1 of 1).

### CODE REVIEW

File: qa/rpc-tests/feature_blocksign.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        signedblock = result["hex"]
+        miner.submitblock(signedblock)
+
+    def mine_blocks(self, num_blocks):
+        for i in range(num_blocks):
+            self.mine_block()
+            self.sync_all()
+
+    def run_test(self):
+        # Check that there's 100 UTXOs on each of the nodes
```
Comment (instagibbs) [inline qa/rpc-tests/feature_blocksign.py]: I don't think these checks are useful in this test. Check the block height instead?

File: qa/rpc-tests/feature_blocksign.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            script += "41"
+            script += k.get_pubkey().hex()
+        script += "{}".format(50+n) # num keys
+        script += "ae" # OP_CHECKMULTISIG
+        return script
+
+    def __init__(self):
+        super().__init__()
+        self.setup_clean_chain = True
+        self.num_nodes = 3
```
Comment (instagibbs) [inline qa/rpc-tests/feature_blocksign.py]: can we factor out "3" for all these places in case we want to expand it?

**Comment by jtimon** [comment]
Fixed nits, moved from 3-of-3 to 7-of-9, that makes it slower (from 11s to 34s in my hardware), but I think that's ok. We can expand further afterwards, but this works now.

File: qa/rpc-tests/feature_blocksign.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        # broadcasts it
+        mineridx = self.nodes[0].getblockcount() % self.num_nodes # assuming in sync
+        miner = self.nodes[mineridx]
+
+        # miner makes a block
+        block = miner.getnewblockhex()
+
+        # collect required_signers signatures
+        sigs = []
+        for i in range(self.required_signers):
```
Comment (instagibbs) [inline qa/rpc-tests/feature_blocksign.py]: We should loop and make sure that combining below-threshhold doesn't give you a block considered valid.

At each inner-section we should combine and check that it's not complete, and cannot move the blockchain forward.

**Comment by stevenroose** [comment]
Hehe, nice to see that thing being even the slightest bit inspiring :) It was mostly me getting familiar with the Python framework :)

Can't wait to join you guys!

File: qa/rpc-tests/feature_blocksign.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            result = miner.combineblocksigs(block, sigs)
+            util.assert_equal(result["complete"], False)
```
Comment (instagibbs) [inline qa/rpc-tests/feature_blocksign.py]: Please also do a submitblock, then check that it didn't move the chain forward. Belts and suspenders!

File: qa/rpc-tests/feature_blocksign.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+# requires Python 3.5
```
Comment (instagibbs) [inline qa/rpc-tests/feature_blocksign.py]: We shouldn't require a specific version of python3.

**Comment by instagibbs** [comment]
utACK https://github.com/ElementsProject/elements/pull/373/commits/f66e94a7120caa4a9a53395853bc837d75aaed6b

@stevenroose See you soon!

**Pull request merged**